modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm

```diff
@@ -316,33 +331,122 @@ sub _write_output {
+	       push @$split_trees, @$subtrees;
   	   }
 	   $gata->store_group($split_trees);
 	   foreach my $tree (@$split_trees) {
+	       $self->_assert_cigar_lines_in_block($tree->modern_genomic_align_block_id);
+	       $self->_assert_tree($tree);
 	       $self->_write_gerp_dataflow($tree->modern_genomic_align_block_id);
 	       
 	   }
        } else {
 	   $gata->store($genomic_align_tree, $skip_left_right_index);
-	   $self->_write_gerp_dataflow($genomic_align_tree->modern_genomic_align_block_id)
+	   $self->_assert_cigar_lines_in_block($genomic_align_tree->modern_genomic_align_block_id);
+	   $self->_assert_tree($genomic_align_tree);
+	   $self->_write_gerp_dataflow($genomic_align_tree->modern_genomic_align_block_id);
        }
    }
-	#DO NOT COMMENT THIS OUT!!! (at least not permenantly). Needed
-	#to clean up after each job otherwise you get files left over from
-	#the previous job.
-    $self->cleanup_worker_temp_directory;
+}
+
+sub _assert_tree {
+    my ($self, $gat) = @_;
+    foreach my $node (@{$gat->get_all_nodes}) {
+        if ($node->is_leaf) {
+            if ($node->genomic_align_group->genome_db->name eq 'ancestral_sequences') {
+                $self->_backup_data_and_throw($gat, sprintf("Node %s is a leaf, but attached to ancestral_sequences", $node->name));
+            }
+        } else {
+            if ($node->genomic_align_group->genome_db->name ne 'ancestral_sequences') {
+                $self->_backup_data_and_throw($gat, sprintf("Node %s is internal, but not attached to ancestral_sequences", $node->name));
+            }
+            if ($node->get_child_count != 2) {
+                $self->_backup_data_and_throw($gat, sprintf("Node %s is not binary", $node->name));
+            }
+        }
+    }
+}
+
+sub _assert_cigar_lines_in_block {
+    my ($self, $gab_id) = @_;
+    my $gat = $self->compara_dba->get_GenomicAlignTreeAdaptor->fetch_by_genomic_align_block_id($gab_id);
+
+    my %lengths;
+    foreach my $genomic_align_node (@{$gat->get_all_nodes}) {
+        foreach my $genomic_align (@{$genomic_align_node->genomic_align_group->get_all_GenomicAligns}) {
+            my $cigar_length = Bio::EnsEMBL::Compara::Utils::Cigars::sequence_length_from_cigar($genomic_align->cigar_line);
+            my $region_length = $genomic_align->dnafrag_end - $genomic_align->dnafrag_start + 1;
+            if ($cigar_length != $region_length) {
+                $self->_backup_data_and_throw($gat, "cigar_line's sequence length ($cigar_length) doesn't match the region length ($region_length)" );
+            }
+            my $aln_length = Bio::EnsEMBL::Compara::Utils::Cigars::alignment_length_from_cigar($genomic_align->cigar_line);
+            $lengths{$aln_length}++;
+        }
+    }
+    if (scalar(keys %lengths) > 1) {
+        $self->_backup_data_and_throw($gat, "Multiple alignment lengths: ". stringify(\%lengths));
+    }
+}
+
+sub _backup_data_and_throw {
```

Comment:
Having these three methods repeated almost to a dot in three runnables seems like a good reason to refactor... Would it be worth it?

Reply:
Agree, but we may be out of time on this one - might have to do it ourselves @JAlvarezJarreta 😉